ksp/src/main/kotlin/com/tschuchort/compiletesting/Ksp.kt

```diff
@@ -0,0 +1,68 @@
+/**
+ * Adds support for KSP (https://goo.gle/ksp).
+ */
+package com.tschuchort.compiletesting
+
+import org.jetbrains.kotlin.ksp.KotlinSymbolProcessingCommandLineProcessor
+import org.jetbrains.kotlin.ksp.KotlinSymbolProcessingComponentRegistrar
+import org.jetbrains.kotlin.ksp.processing.SymbolProcessor
+import java.io.File
+
+private const val KSP_PLUGIN_ID = "org.jetbrains.kotlin.ksp"
+
+private fun KotlinCompilation.initAndGetKspConfig(): KspConfiguration {
+    val config = KspConfiguration(workingDir.resolve("ksp"))
+    if (config.workingDir.exists()) {
+        // already configured, just return
+        return config
+    }
+    config.classesOutDir.mkdirs()
+    config.sourcesOurDir.mkdirs()
+    config.syntheticSources.mkdirs()
+    val kspOptions = listOf(
+        PluginOption(KSP_PLUGIN_ID, "apclasspath", config.syntheticSources.path),
+        PluginOption(KSP_PLUGIN_ID, "classes", config.classesOutDir.path),
+        PluginOption(KSP_PLUGIN_ID, "sources", config.sourcesOurDir.path)
+    )
+    compilerPlugins = compilerPlugins + KotlinSymbolProcessingComponentRegistrar()
+    commandLineProcessors = commandLineProcessors + listOf(KotlinSymbolProcessingCommandLineProcessor())
+    pluginOptions = pluginOptions + kspOptions
+    return config
+}
```

Comment:
What if you have a different instance of `KotlinCompilation` but the same `workingDir`? In that case the function will return at line 17 but `compilerPlugins`, `commandLineProcessors` and `pluginOptions` don't include KSP. Worst case you just have to search the lists if KSP is already added; Should be fast enough.

That being said, it sucks that you have to do this in the first place. The API of `KotlinCompilation` should be changed to make it easier to write addons in a safe way.

Reply:
is that a case where workingDir is re-used? I thought KotlinCompilation ensured there is a new folder for each compilation ?.

Overall in this PR i tried to avoid changing anything in KotlinCompilation. Unfortunately, I'm bound by its public API (since this is a different module) so I had to create these weird solutions :/. LMk if you have a better solution to create a private API between these two gradle modules.

Reply:
It only creates a new temp working directory by default if no other one is given, so it's very much possible that two different instances of `KotlinCompilation` will use the same `workingDir`. IIRC I don't even use that in the tests and let JUnit create the temp folder. Maybe I can think of a nicer solution. I guess searching the list would be at least a little bit better. 

If you have any ideas how to improve the API, let me know and you can definitely change `KotlinCompilation` when it's non-breaking.

Reply:
What if I added two APIs to KotlinCompilation:

```
val metadata : Map<Key, Any> // a place to stash data by extensions
val lifecycleObservers : List<LifecycleObservers> 
interface LifecycleObserver {
    onPreCompile()
    onPostCompile(result : CompilationResult)
}

fun <T> addMetadata(key : Class<T>, value : T)
fun <T> getMetadata(key : Class<T>)
fun addLifecycleObserver(observer : LifecycleObserver)
```
that would provide me enough infra to keep the list of processors and also set it up w/ the plugin right before compilation.

Reply:
here is another prototype w/ instance support:

https://github.com/yigit/kotlin-compile-testing/commit/4017e35f04e0f3ea8a154337af90196c8ba9531d

i just need a way to cleanup the thread local data after compilation for which i'll need a hook from KotlinCompilation.

wdyt? If that implementation + lifecycle hook makes sense to you; i can cleanup and include that in this PR?

Reply:
How would you use the `metadata` and `lifecycleObserver` API? Adding mutable variables to a class like that feels like a hack, but maybe it is the best option without breaking the API. Or we chose the dumb simple way:

```Kotlin
fun KotlinCompilation.addKspPlugin() = this.run {
  // Do whatever needs to be done to initialize KSP support, but in an idempotent way if possible
  // Mkdirs is idempotent
  // Search compilerPlugins list if it already contains KSP plugin, if not, add it
  // Searching the list should be fast enough and if a user calls this multiple times it's their own fault
}  

fun KotlinCompilation.addKspProcessor() = this.run {
  // Do your thing with the bytecode generation
}
```

Even better would be if you could subclass `KspComponentRegistrar` to overwrite the service location so we can pass instances of `SymbolProcessor` directly to the `KspComponentRegistrar` and avoid writing the service declarations to disk. 

Most likely the best breaking solution would involve decoupling the compilation configuration and the function that runs it, using interfaces for the configuration to enable decorator pattern and making the `run` function polymorphic on the configuration type.  

Regarding the bytecode generation: What happens if the `SymbolProcessor` is `final`? Then you can't subclass it in the generator function.

Reply:
`KotlinSymbolProcessingComponentRegistrar` is final so i cannot override it. But that is actually a good idea and I can implement this w/ a custom registrar & extension (luckily AbstractKotlinSymbolProcessingExtension is public but i'm not sure if it was intentional)

```
private class MyAbstractKotlinSymbolProcessingExtension(
        options : KspOptions,
        private val processors : List<SymbolProcessor>
) : AbstractKotlinSymbolProcessingExtension(
        options,
        testMode = false
) {
    override fun loadProcessors() = processors
}

private class MyRegistrar(
        val workingDir: File
) : ComponentRegistrar {
    val processors = mutableListOf<SymbolProcessor>()
    override fun registerProjectComponents(project: MockProject, configuration: CompilerConfiguration) {
        val options = KspOptions.Builder().apply {
            // i should probably clear these
            this.classesOutputDir = workingDir.resolve("out")
            this.sourcesOutputDir = workingDir.resolve("src")
        }.build()
        val registrar = MyAbstractKotlinSymbolProcessingExtension(options, processors)
        AnalysisHandlerExtension.registerExtension(project, registrar)
    }
}
```
this avoids any bytecodegen or even using the KSP plugin itself.
I cannot make use of command line arguments processor but maybe thats OK.

This also has the added benefit of not needing another place to stash values. 
I'll update this PR to use a cleaned up version of the code above and if the KSP team OKs it, i think that might be the best way to go.